### ISSUE TITLE
The "filterMyFiles" method is returning duplicate ID values

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/SlickGrid#3]
For example, typing the letter "m" is showing this in the logs
`[21, 21, 9, 9, 10, 4, 5, 22, 7, 8, 11, 11, 12, 12, 14, 15, 16, 17, 18]`

**Comment by Alex-ley** [comment]
Where are you trying to implement this in your lib? Still in the example-tree-data.html or now in one of the .js files?

Just to make sure, you renamed the ._used prop flag to .__used right? I suspect to fit in with the style of the test of your lib etc?

Could it be that you missed ti make that change in the line ``delete myObj[id]._used``? To also use double underscore? As this line is critical to ensure no duplicates.

Did you commit your refactored code somewhere? Then I can just work directly on that? I’m going to bed now but will take a look tomorrow - I’m also off work :)

**Comment by ghiscoding** [comment]
I changed it to `__used` just to match the other properties used by this filtering (soon sorting), I use 2 underscores to avoid any collisions with any data that might (who knows) have some properties named with 1 underscore, never too careful. My lib is used by few hundred, so you could be surprised by some use case sometime 😮 

It doesn't look like I missed any in the renaming though, all the code is already pushed, you can just refresh your code and work from there. The only thing I added on my side is a console log before `return filteredChildrenAndParents;` to find the issue.

I'm working at merging this in my Slickgrid-Universal lib, it's almost done, starting to clean up the code. Sweet :)

**Comment by ghiscoding** [comment]
@Alex-ley 
I finished merging the code in my lib and thanks to you, it all works as expected... woohoo
The duplicate IDs is still a minor issue, though it still works even then because the `myFilter` just check if the ID is found in the array and duplicate or not that doesn't really matter. You can take a quick look at it if you want but it's really minor.

I've updated my Slickgrid-Universal demo page and that is where the animated gif comes from. 
@Alex-ley I can't thank you enough for your amazing help. 😉 

Next challenge with Tree Data
1. Data Filtering (done, thank you so much)
2. I got single sort working (with recursion and hierarchical dataset) but multi-column sort is not working as expected.
3. Aggregation (I most probably can do it when I sort the data with recursion)

The number 2-3 are just bonuses, not required yet for our project but if we have them then it's just candies treat 🍬. 

Note that I have 2 demos [Example 5](https://ghiscoding.github.io/slickgrid-universal/#/example05) is with a parent/child ref (like the one you helped me with) and the [Example 6](https://ghiscoding.github.io/slickgrid-universal/#/example06) is with a hierarchical dataset (meaning that it already is a tree shape in the JSON dataset)... and they both work! Sweet!
![fC2rDbxcrj](https://user-images.githubusercontent.com/643976/78949222-00c95180-7a99-11ea-9eb8-58245d177018.gif)

**Comment by ghiscoding** [comment]
I might have another use case that I forgot for Filtering. Say we search for "documents", I would like to display the entire content of that folder (all its children), not just the one that matches the word. I'm thinking about a flag that I could enable in my grid options. I didn't check the code yet, hopefully, that won't be too hard to implement, the logic should be:: if we have a parent then include all children under and skip any condition check.

Right now it will only return the parent since that is the only row containing that word. 
![image](https://user-images.githubusercontent.com/643976/78961770-91675800-7ac0-11ea-81c9-e69488d0ac4e.png)

**Comment by Alex-ley** [comment]
Ok, I will spend a couple of hours looking into this all now.

It appears that the duplicates appear with a search of "m" just like you show, but do not appear with a search of "ma" - with "ma" the result is ``[21, 4, 5, 22]``

And to be honest I was never checking the result of just "m" I was always typing at least "ma" (and other 2 letter combinations) before checking the result etc.

The other use case for filtering - i.e. to show all children of a folder, might require a bit of a different approach as so far we were only ever searching "up" the tree to parent and then parent of the parent etc. and never "down" the tree to child and then child of child etc.

I will think about how to achieve this and let you know my thoughts etc. :)

**Comment by ghiscoding** [comment]
Thanks again, I'll take all the help I can have but it's Easter Holiday after all, at least here in Canada so relax if you can. But it's not same since I can't go visit any family relatives... might resolve in working anyway haha. Cheers

**Comment by Alex-ley** [comment]
Yeah, it's also a 4 day weekend in Germany but without being able to visit family or friends it leaves still a lot of time for coding! I was reading & relaxing most of the day then went for a walk and did some exercise in the house then after that attempted to work this out :)

I have a working idea here in codePen (https://codepen.io/Alexander9111/pen/VwvYpaX) but again it's hard to integrate into the example-tree-data.html.

I think the way I have done it this time (generate a real hierarchical tree object and then search that recursively) means that it would be a better idea to generate this treeObj upfront and not each time the ``filterMyFiles()``  function is called.

I will try to figure out where to put that and how to do it tomorrow :)

**Comment by ghiscoding** [comment]
@Alex-ley 
Hello Alex, long time since we talked, I'm slowly starting to look at this again. Apart from the Aggregator that I said I was going to look into (planning on doing so in coming few days/weeks), was there anything else? I see that I could maybe add the Sorting in your `example-tree-data.html`, I wonder if it will work or if the aggregators will give problem with the sort? Let me know if there's anything else, I'd be happy to merge this code in the main repo (6pac/SlickGrid). Thanks again for all of this, I'm sure many people will soon benefit from your work :)

**Comment by Alex-ley** [comment]
@ghiscoding - sorry for the slow reply! Erm, yeah I think aggregators and filtering would be everything, right?

How’s it going so far? I’ll pull your commits tomorrow and take a look and how it’s working. We could chat tomorrow or next weekend etc. if it would be more efficient?

Hope you’re doing well!

**Comment by ghiscoding** [comment]
Hey Alex, no worries I'm slowing down too a bit, it's summer after all, gotta enjoy it somehow... even if it's close to home hehe. 
I didn't look at the Aggregators yet, the PR only adds the Sorting but even that is not complete, and just for reference sake, here's the [commit](https://github.com/ghiscoding/slickgrid-universal/commit/0861295779dcb6012635e3e7deb6e753ba9bdc81#diff-bede61524fc5bba92add1cf789b1c80e) in which I tried adding built-in Aggregators which I later removed. I plan to review that code to give it another go at it. 

BTW I'm meeting a friend tomorrow, 1st meeting in many months, so I won't be home. However I'm working only 2 days next week then I'm off for a 1.5 week so we could maybe talk during that period.

**Comment by Alex-ley** [comment]
Hey @ghiscoding sounds good! Hope the weekend was enjoyable. Let me know what day(s) might work in the next couple weeks? I would be available most days around 6pm CET and Fridays I’m not working the next couple months 👌

**Comment by ghiscoding** [comment]
Oh ok great, perhaps this coming Saturday or Sunday, that would be around noon for me I think (I think we have 12 hours difference) and that would be ok on my side. I just shipped another version of Angular-Slickgrid an hour ago, so that's another thing out of the way :)
